### ISSUE TITLE
WebGPU-based renderer for the editor

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#221145]
We're finally starting to look at implementing a WebGPU-based rendering in monaco, similar to what xterm.js uses. This issue is used to track all the work which is expected to take several months.

Project: https://github.com/orgs/microsoft/projects/1367/views/1

## Related issues

Here are some historical links that might be useful:

- Zero latency typing issue which investigated typing latency in monaco. This lead to some improvements but eventually came to the conclusion that the biggest win would be doing rendering manually using our own shaders https://github.com/microsoft/vscode/issues/27378
- Old exploration issue https://github.com/microsoft/vscode/issues/162445
- Internal issue describing my 2022 prototype which is based on xterm.js' webgl renderer https://github.com/microsoft/vscode-internalbacklog/issues/3157
  - The branch https://github.com/microsoft/vscode/tree/tyriar/webgl_monaco
- Internal issue describing my 2024 prototype which is based on a game renderer I was experimenting with in my spare time  https://github.com/microsoft/vscode-internalbacklog/issues/4906
  - The branch https://github.com/microsoft/vscode/tree/tyriar/gpu_exploration

---

Below copied from https://github.com/microsoft/vscode-internalbacklog/issues/4906

## GPU-based rendering

branch: [tyriar/gpu_exploration](https://github.com/microsoft/vscode/tree/tyriar/gpu_exploration)

### How GPU rendering works

It works by assembling array buffers which represent commands to run on the GPU, these are filled on the CPU with information like the texture to use (chracter, fg, bg), location, offset, etc. xterm.js for example allocates a cols x rows array buffer that represents the viewport only and updates it on every frame where the viewport changes.

There are 2 types of shaders:

- Vertex shader - This is run for every vertex (4 vertices per quad) and is used to transform the vertices into screen space.
- Fragment shader - This is run for every pixel in the quad and is used to determine the color of the pixel.

### How the prototype works

The WebGPU prototype works by pre-allocating a buffer that represents up to 3000 lines in a file with a maximum column length of 200. The buffers\* are lazily filled in based on what's the viewport. Meaning once a line is loaded, it doesn't need to be modified again. I think it updates more aggressively currently than needed due to my lack of knowledge around finding dirty lines in Monaco.

```wgsl
@vertex fn vs(
	vert: Vertex,
	@builtin(instance_index) instanceIndex: u32,
	@builtin(vertex_index) vertexIndex : u32
) -> VSOutput {
	let dynamicUnitInfo = dynamicUnitInfoStructs[instanceIndex];
	let spriteInfo = spriteInfo[u32(dynamicUnitInfo.textureIndex)];

	var vsOut: VSOutput;
	// Multiple vert.position by 2,-2 to get it into clipspace which ranged from -1 to 1
	vsOut.position = vec4f(
		(((vert.position * vec2f(2, -2)) / uniforms.canvasDimensions)) * spriteInfo.size + dynamicUnitInfo.position + ((spriteInfo.origin * vec2f(2, -2)) / uniforms.canvasDimensions) + ((scrollOffset.offset * 2) / uniforms.canvasDimensions),
		0.0,
		1.0
	);

	// Textures are flipped from natural direction on the y-axis, so flip it back
	vsOut.texcoord = vert.position;
	vsOut.texcoord = (
		// Sprite offset (0-1)
		(spriteInfo.position / textureInfoUniform.spriteSheetSize) +
		// Sprite coordinate (0-1)
		(vsOut.texcoord * (spriteInfo.size / textureInfoUniform.spriteSheetSize))
	);

	return vsOut;
}
```

```wgsl
@fragment fn fs(vsOut: VSOutput) -> @location(0) vec4f {
	return textureSample(ourTexture, ourSampler, vsOut.texcoord);
}
```

### Texture atlas

Glyphs are rendered on the CPU using the browser's canvas 2d context to draw the characters into a texture atlas. The texture atlas can have multiple pages, this is an optimization problem as uploading images is _relative_ expensive. xterm.js creates multiple small texture atlas pages, allocates using a shelf allocator and eventually merged them into larger immutable pages as they're more expensive to upload.

Currently the prototype uses a single large texture atlas page, but it warms it up in idle callbacks for the current font and all theme token colors in the background (using the `TaskQueue` xterm.js util).

![image](https://github.com/user-attachments/assets/6e12c0d1-a7d0-4451-a2d6-6632a1f80027)

### Memory usage

```
text_data_buffer: [wgslX, wgslY, textureIndex, ...]

texture_atlas_buffer: [positionX, positionY, sizeX, sizeY, offsetX, offsetY, ...]

textureIndex in text_data_buffer maps to texture_atlas_buffer[textureIndex * 6]
```

In the above, each text_data_buffer cell is 12 bytes (3x 32-bit floats), so 3000x200 would be:

```
3000 * 200 * 12 = 7.2MB
```

This is pretty insignificant for a modern GPU.

\* Double buffering is used as the GPU locks array buffers until it's done with it.

### Scrolling

The prototype currently scrolls extremely smoothly as at most a viewport worth of data is filled but often no viewport data will change. Then we just need to update the scroll offset so the shadow knows which cells to render.

### Input

So far, the above is highly optimized for readonly scrolling. For input/file changes there are a few cases we need to target. We essentially want to get these updates to take as little CPU time as possible, even if that means leaving stale and no-longer referenced data in the fixed buffers.

**Adding new lines or deleting lines**

This could be supported by uploading a map whose job is to map line numbers with the index in the fixed buffer:

![image](https://github.com/user-attachments/assets/8b6badfd-f793-4a79-b12d-c0e6f94d823e)

That way we only need to update indexes, not the whole line data.

**Inserting characters**

Simple O(n) solution is to just update the entire line. We could do tricks to make this faster but it might not be worth the effort if line length is fixed.

### Fixed buffers and long lines

My plan for how the characters will be send to the GPU is to have 1 or more fixed width buffers (eg. 80, 200?) with maps that point to indexes dynamically as described in the input section and then another more dynamic buffer which supports lines of arbitrary length. This dynamic buffer will be a little less optimized as it's the edge case when coding. The fixed buffers could also be dynamically allocated based on the file to save some memory.

### Other things we could do

- Sub-pixel glyphs for smoother flow - eg. render characters at 4x the width and support offsetting the character every 0.25px.
- Proportional font support isn't in xterm.js but it's possible without too much effort, we will need to support this anyway if we want to render widths just like the DOM renderer. The main thing this requires is some way of getting the width of the glyphs and the offset of each character in a line. Again this is an optimization problem of getting and updating this width/offst data as fast as possible.
- I believe SPAA is possible to do on the GPU using grayscale textures.
- Custom glyphs are supported in the terminal which allows pixel perfect box drawing characters for example like `┌───┘`. Whether this looks good in monaco is up to the font settings. Letter spacing and line height will always mess with these
- Texture atlas glyphs could be first drawn to a very small page and then packed more efficiently into a longer-term page in an idle callback or worker.
- Texture atlas pages could be cached to disk
- Canvas sharing - To optimize notebooks in particular we could have a shared canvas for all editors and tell the renderer that it owns a certain bounding box

### Test results

These were done on terminalInstance.ts. Particularly slow frames of the test are showed.

The `tyriar/gpu_exploration` tests disabled all dom rendering (lines, sticky scroll, etc.) to get an idea of how fast things **could be** without needed to perform layouts on each frame. It's safe to assume that rendering other components would be less than or equal to the time of the most complex component (minimap is similar, but could potentially share data as well).

**Scroll to top command**

M2 Pro Macbook main

![image](https://github.com/user-attachments/assets/4135a433-e019-474e-96bc-69969bf46cf7)

M2 Pro Macbook tyriar/gpu_exploration (all dom rendering disabled)

![image](https://github.com/user-attachments/assets/72f531ab-aef8-453e-b4e0-e44b1f981c63)

Windows gaming PC main

![image](https://github.com/user-attachments/assets/f96b0915-fb39-4203-9e55-d45170d99316)

Windows gaming PC tyriar/gpu_exploration (all dom rendering disabled)

![image](https://github.com/user-attachments/assets/79abf70f-5265-480d-aa31-c2c46f12c11e)

**Scrolling with small text on a huge viewport**

fontSize 6, zoomLevel -4

M2 Pro Macbook main

![image](https://github.com/user-attachments/assets/94b30603-c59d-4381-9917-4af0f1518795)

M2 Pro Macbook tyriar/gpu_exploration (all dom rendering disabled)

![image](https://github.com/user-attachments/assets/187abe65-f6e9-487c-b36e-0ba220ae017b)

Windows gaming PC main

![image](https://github.com/user-attachments/assets/63c1a618-0ab3-4cca-8750-f2cf02a16173)

Windows gaming PC tyriar/gpu_exploration (all dom rendering disabled)

![image](https://github.com/user-attachments/assets/4bdc4f93-f018-4830-91e0-58bea6f6b383)

**Very long line**

Long lines aren't supported in the gpu renderer currently

#### Shaders run in parallel to microtasks and layout

The sample below from the Windows scroll to top test above demonstrates how the shaders execute in parallel with layout, as opposed to all after layout.

Before:

![image](https://github.com/user-attachments/assets/0ed78860-caf5-4b18-afbc-e9194a64bb60)

After:

![image](https://github.com/user-attachments/assets/a6750842-6ddc-4d99-9638-e58d5f0c632d)

---

Harfbuzz shaping engine is used by lots of programs including Chromium to determine various things about text rendering. This might be needed for good RTL/ligature/grapheme rendering.

**Comment by IllusionMH** [comment]
Would it be possible to re-upload images to this repo, as they aren't available for regular users now, or they are on a need-to-know basis?

**Comment by Tyriar** [comment]
@IllusionMH opps, doing it now 👍

**Comment by Tyriar** [comment]
Update on my end for last week. WIP branch https://github.com/microsoft/vscode/pull/225413

## General

- Rendering is fixed up when switching editors and resizing canvas
- Correct background color is drawn (instead of black)
  ![image](https://github.com/user-attachments/assets/a38dddb6-7298-4bd2-9335-eb6d6ae73970)
- Bunch of general clean up and refactors. In particular improving of variable/constant names and simplifying of the main webgpu code
- Set up a `GPULifecycle` namespace with helpers that return `IDisposable`s
- Sorted out some high level lifecycle/leak issues
- The `GlyphRasterizer` is now owned by `GpuViewLayerRenderer`. The idea here is that the texture atlas is shared across all editors, but different editors could have different font sizes so it's owned by the editor so multiple font sizes can be rendered (WIP, sizes aren't tracked in atlas keys yet).

## Rasterization

- Bold and italic text is now rendered
  ![image](https://github.com/user-attachments/assets/5ed9c943-e794-47c5-9639-4671e50d3eb1)
- Less garbage collection by reusing hot objects

## Texture atlas

- Multiple texture atlas pages are now addressable. There is no overflow logic yet, but glyphs are distributed whether they are alphabet chars in order to test multiple pages
- Glyphs are uniquely identified and stored by their metadata instead of just their foreground color
https://github.com/microsoft/vscode/blob/3cfe905cb8035edf2b824efc865f9d0d0febca7e/src/vs/editor/browser/view/gpu/atlas/textureAtlasPage.ts#L81-L84
- Texture atlas debug commands
  ![image](https://github.com/user-attachments/assets/4d76c5ac-808d-4599-8e44-e45f323cece3)
  - Saving texture atlas pages:
    ![image](https://github.com/user-attachments/assets/354926db-79d0-44f6-9422-ba10164403a8)
  - Logging texture atlas page stats: 
    ![image](https://github.com/user-attachments/assets/72f6e61e-628b-4866-b092-4d22d2748685)
- Some basic unit tests for atlas and allocators
- Only the used portion of the atlas texture is uploaded, speeding up render time when there are new glyphs significantly, especially on initial render (~15ms -> ~2ms)

## Explorations

- Explored approach for rendering of view parts, starting with the ruler.
  - I first tried to do multiple passes with a separate shader but it's more complicated than I initially thought and requires juggling some textures. Additionally, order of render passes and having them all run every time would be required for this.
  - I think the right approach here at least for simple view parts is to allow parts to register shapes into some render pass/command encoder object. This would make the ruler component even simpler than the DOM-based one as they would basically just register some fixed rectangles/lines and then refresh it when the setting changes.
- Explored the "scratch page" idea for the texture atlas.
  - This needed more logic in the shader than expected. Uploading only relevant parts of the page texture was a big win that makes this no longer needed.

**Comment by faheemstepsharp** [comment]
Hope this become default soon.

**Comment by Tyriar** [comment]
@faheemstepsharp I suspect it's going to be a long road to be the default (6 months, 1 year+?). We did eventually switch the terminal to default to GPU rendering, it'll be really bad if we ship an editor that breaks text rendering though.

**Comment by Tyriar** [comment]
Update for @hediet and myself for last week. WIP PR #225413

## Architecture

We came up with a better approach for where to stick the implementation. GPU parts are now regular "view parts" instead of being more tightly tied to the view.

https://github.com/microsoft/vscode/blob/bd21f3c8f268b741edac137fc109d75a555c0541/src/vs/editor/browser/view.ts#L161

A new `ViewGpuContext` contains all objects needed for managing rendering to the frame (canvas element, GpuContext, command encoder, etc.). This is owned by `View` and will be injected to every GPU-related view part, similar to `ViewContext`.

https://github.com/microsoft/vscode/blob/bd21f3c8f268b741edac137fc109d75a555c0541/src/vs/editor/browser/view.ts#L146-L148

❔ The term "context" is becoming a little overloaded (ViewContext, ViewGpuContext, GPUContext). Maybe there's a better name for `ViewGpuContext`?

## Drawing shapes

Built out the `ObjectCollectionBuffer` data structure which allows creating type-safe objects that get encoded into a `Float32Array` which will be used to draw shapes via the `ViewGpuContext` interface. This will allows view parts to easily add, remove and change sections of the `Float32Array` in a fairly performant way without needing to deal with the actual buffer. Done right I think this should make the implementation of simple view parts like `rulers` to be even simpler than the DOM-based counterpart.

https://github.com/microsoft/vscode/blob/bd21f3c8f268b741edac137fc109d75a555c0541/src/vs/editor/test/browser/view/gpu/objectCollectionBuffer.test.ts#L42-L54

This object isn't hooked up yet, just the data structure and tests are mostly done.

## General

- Lots of cleaning up of interfaces, adding jsdoc, etc.
    - Removed chars/tokenFg from the allocator interface, to makes it more clear that all an allocator's job is to take a rasterized glyph, put it into an atlas and track the usage.
- Fixed metadata key to _properly_ remove metadata that doesn't affect the glyph's rendering.
- Fixed "null cells" rendering random characters to the middle of the renderer. This was happening because zeroed out sections of the buffer were all pointing at the first glyph of the first atlas page.
- The canvas is sized to fit `.overflow-guard`. This is probably the final size and position of the canvas.
    ![image](https://github.com/user-attachments/assets/f2bb741b-b4ca-40c2-9ce7-546f5429eb66)
- Added viewport offset which now renders the characters in approximately the right position (when dpr=1 at least). The top and the bottom lines in this picture show the gpu rendering overlaid on top of the DOM rendering.
    ![image](https://github.com/user-attachments/assets/96653b44-d248-437d-93bc-ffec57c2a6a9)
- Added #regions and organized the webgpu init code a little better.
    ![image](https://github.com/user-attachments/assets/08210920-8aac-49e5-86ad-f7b98ed2fef7)
- Added a hidden setting to enable the GPU renderer so we can merge the code with minimal impact on default rendering.

## Texture atlas

- Basic page overflow logic is done; when a page is filled it will start adding glyphs to a second page. Only 2 pages are currently supported in the shader though.
- Handle edge cases around glyphs too large for slab or page.
- Reduced search time for glyph's page to O(1) https://github.com/microsoft/vscode/pull/225413/commits/740ba1ce332ea24b51baef034c3ccfa926d3790f
- More tests!

## Debugging

- New draw glyph command that will draw a single glyph.
    ![image](https://github.com/user-attachments/assets/84420267-48fe-4529-877f-73931a4952be)
- Consolidated all gpu-related debug command into a single `Developer: Debug Editor GPU Renderer` command that brings up a picker. This will let us create many debug-related commands without spamming the regular command palette.
    ![image](https://github.com/user-attachments/assets/5f20aa3f-8f83-44f5-af44-3d8dd740d137)

**Comment by vincentdchan** [comment]
This may be a silly question but how do you draw glyphs on WebGPU? Are you drawing the glyph map with canvas or render the font manually?

**Comment by Tyriar** [comment]
@vincentdchan fonts are rasterized to a texture atlas using a 2d canvas context (mostly on CPU), then the texture atlas pages are uploaded and drawn by the shader where each cell is a texture mapped to 2 triangles. So we're leveraging the browser's font rendering stack and can avoid getting into that.

**Comment by vincentdchan** [comment]
@Tyriar Great! I am implementing a canvas in WebGPU but AFAIK, 2D canvas doesn't provide API to detect `font ligatures`. One  approach I used is reading from the font file. Do you handle `font ligatures`?

**Comment by Tyriar** [comment]
@vincentdchan it does not, ligatures are still not supported in the terminal but they're close. The approach uses there is to parse out the ligatures from the font file and then draw the character sequences to the canvas together via a "character joiner" concept, such that they are rendered as ligatures: https://github.com/xtermjs/xterm.js/blob/f186475ec9375256d304fb4563160e2cd3fef291/addons/addon-ligatures/src/LigaturesAddon.ts#L35

There's also a set of "fallback" ligatures which makes it mostly work when font access isn't granted on the web.

**Comment by Tyriar** [comment]
New public GH project to track this work: https://github.com/orgs/microsoft/projects/1367

**Comment by Tyriar** [comment]
I didn't have too much time last week to work on this but here's what we got done:

- Multiple font sizes and font families now work at the same time
  ![Image](https://github.com/user-attachments/assets/3a6a5ce9-687a-4201-9503-b83564fe5c4a)
- Correct tab offset calculation based on `editor.tabSize` (see screenshot above)
- Texture atlas is cleared when the client is zoomed in or out, so text no longer shows as wrong size
- Implemented a simple version of `canRender` and respected that in the GPU renderer, so lines won't get double rendered anymore
- Start WIP for the "rectangle renderer" which will be the backend for drawing cursors, current line height, rulers, etc. https://github.com/microsoft/vscode/pull/228632

**Comment by Tyriar** [comment]
I was busy experimenting with the new Copilot API this month, will start back up on this effort in November.

**Comment by Tyriar** [comment]
Update for the last 2 weeks:

The main goal was to start self-hosting and I've accomplished that as I started using it 100% of the time on Thursday since selection and mouse events are finally working! It's still rough around the edges but it's definitely usable especially starting from the upcoming Insiders release on Monday.

If you'd like to try it out the setting is currently `"editor.experimentalGpuAcceleration": "on"`, it will fade and say it's unrecognized by design for now as I don't want users accidentally stumbling across it so early on. Note that the performance benefits probably won't show just yet, I expect a big bump in performance when [line numbers ](https://github.com/microsoft/vscode/issues/233988) and [bracket pair colorization](https://github.com/microsoft/vscode/issues/233989) is handled on the GPU side. Right now any decoration will fallback to DOM, so turning off `editor.bracketPairColorization.enabled` will likely make most lines render on the GPU (if the codebase is mostly ASCII).

These are the issues currently being tracked: https://github.com/microsoft/vscode/labels/editor-gpu, it might not be worth creating issues just yet as it doesn't take very long to run into issues so a lot of them are obvious and already tracked. Commenting here about any problems if you hit them would probably be best for now.

## Major changes

- Add an overlay to indicate GPU lines which shows all the time when the setting is on currently (temporary always on) https://github.com/microsoft/vscode/pull/233932
  ![Image](https://github.com/user-attachments/assets/ed77a52a-29a8-4677-89a0-f51bf20315fa)
- Get DOM-based selections working on GPU lines https://github.com/microsoft/vscode/pull/233769, https://github.com/microsoft/vscode/pull/233923
  ![Image](https://github.com/user-attachments/assets/77d82dcd-2ef0-46a2-8251-434ee6a6b31d)
- Get mouse events working on GPU lines https://github.com/microsoft/vscode/pull/229172

## Performance

- Add caching to the line data to avoid needing to send so much data to the GPU https://github.com/microsoft/vscode/pull/233947
- Only update scroll GPU buffer on the scroll event https://github.com/microsoft/vscode/pull/233796

## Bug fixes / misc

- Re-render when a bunch of events happen https://github.com/microsoft/vscode/pull/232975, https://github.com/microsoft/vscode/pull/233674, https://github.com/microsoft/vscode/pull/233781, https://github.com/microsoft/vscode/pull/233837, https://github.com/microsoft/vscode/pull/233934
- Log when uploading a 3rd+ texture atlas page to make it obvious why things don't look right https://github.com/microsoft/vscode/pull/232986
- Fix the editor being blank sometimes when an editor is opened https://github.com/microsoft/vscode/pull/232998
- Fix left boundary tab rendering https://github.com/microsoft/vscode/pull/233666
  ![Image](https://github.com/user-attachments/assets/e6d7a034-062d-446b-b55f-e39b3b5feaca)
- Adjust content after lines are deleted https://github.com/microsoft/vscode/pull/233680
- Allow falling back to DOM for lines 3000+ (temporary) https://github.com/microsoft/vscode/pull/233696
  ![Image](https://github.com/user-attachments/assets/91c6c401-7b1e-4f71-a226-b4fa87367c5d)
- Clear space and tab cells on enter and don't use the tab offset in the buffer https://github.com/microsoft/vscode/pull/233701/
- Move WGSL coordinate calculation inside WGSL, meaning less CPU work and lines are more cacheable https://github.com/microsoft/vscode/pull/233789
- Fix an issue where glyphs sometimes missed the last row and column of pixels https://github.com/microsoft/vscode/pull/233918
- Fix an issue where negative zoom would break everything https://github.com/microsoft/vscode/pull/233918

**Comment by Tyriar** [comment]
Update for the week:

- Inline decorations which were one of the big unknowns now have partial support and we have a way forward. Currently in main decorations that change `color` only like bracket pair colorization should work. See https://github.com/microsoft/vscode/pull/234127 for details. This too a lot of my time to prototype and then polish the large PR.
  ![Image](https://github.com/user-attachments/assets/05d16e1c-42bf-4e5a-86d2-34c620182dc1)
- The GPU mark is now inverted, showing for DOM-rendered lines https://github.com/microsoft/vscode/pull/233995. Hovering it will show the reason(s) why it can't be rendered:
  ![Image](https://github.com/user-attachments/assets/2885db49-c98f-429d-a75b-653ce118df49)
  ![Image](https://github.com/user-attachments/assets/3c9596d4-1c09-4b8e-b416-b2d544777cc3)
- The line numbers region is now clipped so we should not be rendering lines underneath it when scrolled to the right https://github.com/microsoft/vscode/pull/234123. Also should not show underneath the scroll bar https://github.com/microsoft/vscode/pull/234125
  ![Image](https://github.com/user-attachments/assets/2ac6200c-e1e6-4ebd-bab6-be77073a64a3)
- Fixing up several problems around render data caching https://github.com/microsoft/vscode/pull/234104, https://github.com/microsoft/vscode/pull/234107, https://github.com/microsoft/vscode/pull/234264

**Comment by AurevoirXavier** [comment]
@Tyriar, thanks for your work and updates! However, I believe most people won’t fully comprehend the intricacies of these works. If you could also provide a rough estimate of the progress (even a very rough one is sufficient), that would be fantastic! We eagerly anticipate the availability of this feature in the vscode-prerelease.

**Comment by Tyriar** [comment]
Last week was a short one due to holidays in the US. Mostly just fixing some of the bigger self hosting problems.

- Improve rendering to the alphabetical baseline so lines don't jump around when switching between GPU and DOM lines (tested only on devicePixelRatio 1 and 2) https://github.com/microsoft/vscode/pull/234601
- Fix various problems with handling view events https://github.com/microsoft/vscode/pull/234613, https://github.com/microsoft/vscode/pull/234618
- Fix mid-line tabs https://github.com/microsoft/vscode/pull/234661
- Fix text being wrong size after zooming https://github.com/microsoft/vscode/pull/234698
- Decorations
  - Disable all non-regular inline decorations for now https://github.com/microsoft/vscode/pull/234579
  - Fix bracket pair colorization inline decorations not working sometimes https://github.com/microsoft/vscode/pull/234697

**Comment by MoonE** [comment]
I added the config to enable GPU rendering but all lines that should have been rendered on the GPU disappear.
Is this expected? I'm not sure chromium on linux supports webgpu, but an error message or refusing to enable it would be nice to have.
![Image](https://github.com/user-attachments/assets/d94ac812-5948-4f9d-84d6-b8bc682caff5)

Thanks for your work and looking forward to this.

Version: 1.96.0-insider
Commit: 9e67a1dad5d030b5e2d4a1bb70f12c82287135db
Date: 2024-12-09T16:08:03.105Z
Electron: 32.2.6
ElectronBuildId: 10629634
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Linux x64 6.11.0-9-generic (Ubuntu 24.10)

**Comment by MoonE** [comment]
> you need to reload after doing changing the setting.

I did, the text disappears after that.

> but it's likely related to Linux, especially if you're in a VM.

Not in a VM, so it probably is a Linux issue, seems like my hardware is on a chromium blocklist.

`Developer: Show GPU Info` shows this:
> WebGPU has been disabled via blocklist or the command line


|GPU config|value|
|------------|------------|
| 2d_canvas | enabled |
| canvas_oop_rasterization | enabled_on |
| direct_rendering_display_compositor | disabled_off_ok |
| gpu_compositing | enabled |
| multiple_raster_threads | enabled_on |
| opengl | enabled_on |
| rasterization | enabled |
| raw_draw | disabled_off_ok |
| skia_graphite | disabled_off |
| video_decode | enabled |
| video_encode | disabled_software |
| vulkan | disabled_off |
| webgl | enabled |
| webgl2 | enabled |
| webgpu | disabled_off |
| webnn | disabled_off |

Thanks for taking your time to respond.

**Comment by KaeLL** [comment]
Same as MoonE :/


![Image](https://github.com/user-attachments/assets/8bf96c80-d069-41cb-8696-09653ee92599)
![Image](https://github.com/user-attachments/assets/5a2dcca6-a719-4105-b4e6-e1f1a1f12fc5)
![Image](https://github.com/user-attachments/assets/e078188b-01d8-4cae-913f-2fefa7fee351)

**Comment by Tyriar** [comment]
@KaeLL is MoonE a Linux distro? The info in that screenshot is handy, how did you get it?

**Comment by Tyriar** [comment]
@KaeLL the info in that screenshot is handy, how did you get it?

**Comment by KaeLL** [comment]
Same as MoonE
Ctrl + Shift + P -> `Developer: Show GPU Info`

**Comment by Tyriar** [comment]
@KaeLL well that's handy 😅

**Comment by KaeLL** [comment]
You can forcibly enable it by launching `code` with `--enable-features=Vulkan --enable-unsafe-webgpu` Chromium options.
```
Process Argv:     --enable-features=Vulkan --enable-unsafe-webgpu --crash-reporter-id 97fdaa1d-a108-4830-98a6-1d9208b3ecca
GPU Status:       2d_canvas:                              enabled
                  canvas_oop_rasterization:               enabled_on
                  direct_rendering_display_compositor:    disabled_off_ok
                  gpu_compositing:                        enabled
                  multiple_raster_threads:                enabled_on
                  opengl:                                 enabled_on
                  rasterization:                          enabled
                  raw_draw:                               disabled_off_ok
                  skia_graphite:                          disabled_off
                  video_decode:                           enabled
                  video_encode:                           disabled_software
                  vulkan:                                 enabled_on
                  webgl:                                  enabled
                  webgl2:                                 enabled
                  webgpu:                                 enabled
                  webnn:                                  disabled_off
```
I tested that on Chrome 131.0.6778.139 and it explicitly complains after every launch
![Image](https://github.com/user-attachments/assets/6e3740b5-df8f-4ba8-8b65-03776a8200ac)
which, if it serves as a reference to the likely older version of Chromium bundled with Electron, I guess it should be taken more seriously?!

**Comment by Tyriar** [comment]
Short one again as we just came back from the break:

- Rulers are no longer rendered in line number area
- Created a generic solution to more styles needed to be supported by decorations, so far we support color and font-weight (new) https://github.com/microsoft/vscode/pull/237677. See these lines are no longer falling back to DOM:

    ![Image](https://github.com/user-attachments/assets/71acf736-1285-4339-9f56-72ba4466713d)

**Comment by IllusionMH** [comment]
@Tyriar are there plans to improve support for font-weights that are not normal or bold?
Currently Cascadia Code has sweet spot at `"editor.fontWeight": 350` (my whole motivation to add number values) but with `"editor.experimentalGpuAcceleration": "on"` lines rendered with WebGPU looks much bolder

**Comment by Tyriar** [comment]
@IllusionMH looking at the code I thought only normal and bold were supported by decorations currently but I did not test this to make sure.

Also I don't think font weight via that setting works at all yet which is tracked in https://github.com/microsoft/vscode/issues/227122, only token-based weight and decorations after the recent change should work. Both of which are just bold or not bold. The goal is to make it look identical, so if DOM-based render does support numbers there I'll need to make sure that's carried over in the GPU one too. That would be done in https://github.com/microsoft/vscode/issues/227122 though

**Comment by IllusionMH** [comment]
Oh, sorry my bad. My comment was about font weight in editor as I haven't found #227122 in open issues and thought it was repurposed/closed. 🤦

**Comment by KaeLL** [comment]
@Tyriar How exactly is this going to be achieved? Wouldn't WebGPU need to be enabled and stable on Chromium/Electron/Whatever for you to start working on this?

**Comment by Tyriar** [comment]
@KaeLL it is enabled on some browsers by default now. I see Chrome doesn't enable by default on Linux yet but Edge does: https://caniuse.com/webgpu. Whether Electron enables it is what's important for us, and we can pass on flags to Electron if needed.

**Comment by MoonE** [comment]
Could be your hardware.

![Image](https://github.com/user-attachments/assets/a60960fb-4a2c-4750-9d7c-3bdeeefc68ab)

**Comment by Tyriar** [comment]
Yes, we still respect Chromium's defaults here since they know a lot more about the hardware/webgpu compat than we do.

**Comment by Tyriar** [comment]
Update for last week:

**New stuff**

- Expose the `editor.experimentalGpuAcceleration` setting to users https://github.com/microsoft/vscode/issues/238175
- Support opacity inline decorations (like unused variable) https://github.com/microsoft/vscode/pull/237822
    ![Image](https://github.com/user-attachments/assets/63a5fa0d-aa28-4e6c-82fa-cd559b568ef0)
- Wrapped lines are now supported and will no longer fallback to DOM unless the line is > 200 columns https://github.com/microsoft/vscode/issues/238138
    ![Image](https://github.com/user-attachments/assets/2f8686ca-d8c3-4bbb-90e3-192fb36c2e49)
- Support emoji https://github.com/microsoft/vscode/pull/238059, https://github.com/microsoft/vscode/pull/238129
    ![Image](https://github.com/user-attachments/assets/5e32e5ee-c787-4a46-b1e7-aa4360e095e6)
- Get proportional fonts and non-standard character widths working https://github.com/microsoft/vscode/pull/238132/
    ![Image](https://github.com/user-attachments/assets/ad9b9f2c-3ee4-4d6b-ac30-23d560f8e1c5)
- (Mostly) support sub-pixel antialiasing https://github.com/microsoft/vscode/pull/238173. It only works for token metadata currently, not decorations like find:
    ![Image](https://github.com/user-attachments/assets/715a7068-0121-490b-9a67-e4d483581c1e)

**Bug fixes/perf/misc internal**

- Fixed rendering/fallback breaking in files with > 3000 lines https://github.com/microsoft/vscode/pull/237824
- Introduce NKeyMap data structure https://github.com/microsoft/vscode/pull/237825
- Reduce the time for GPU lines to show up. These should now show as soon as DOM lines, except for maybe with the first editor https://github.com/microsoft/vscode/pull/238158
- Fix tough bug where lines would double render after toggling word wrap https://github.com/microsoft/vscode/pull/238137

**Comment by IllusionMH** [comment]
I wonder if using GPU renderer it's possible to resolve issue with emojis width, that break alignment for next characters (e.g. #172349)

![Image](https://github.com/user-attachments/assets/8f219f21-ede5-42d6-aa96-3de4b4446cfb)

**Comment by Tyriar** [comment]
@IllusionMH we could definitely more easily force it to the grid and treat it as a wide monospace character. To date the GPU renderer tries to match exactly all the dimensions the DOM renderer expects, so not sure what other challenges would come up by trying to do that. With how often lines fall back, like with `"git.blame.editorDecoration.enabled": true` the current line always does, this would look really odd for example. I'm not sure we'll ever be at a place where all possible inline decorations are supported since they can use arbitrary CSS atm.

I've also hit this and found it a little weird with emoji and CJK chars. Maybe something we could consider doing later as an opt-in setting?

**Comment by aurexav** [comment]
Feedback from VSCode Insiders:

Scrolling through a long comment causes noticeable lag, whereas scrolling through code remains smooth. I’m not sure why this happens.

**Comment by Tyriar** [comment]
@aurexav thanks, I created https://github.com/microsoft/vscode/issues/238849 for 1. For 2, could you create an issue with the code that caused that and what you did if anything to make it show? I'm not too familiar with Rust.

**Comment by Tyriar** [comment]
Update:

- New viewport rendering strategy which means we can not render files with many lines, as well as lines up to 2000 characters before falling back https://github.com/microsoft/vscode/pull/238487
    ![Image](https://github.com/user-attachments/assets/9ab89040-d4eb-4cc3-8d3e-6ac9932b883a)
- Added sub-pixel offset rendering of glyphs https://github.com/microsoft/vscode/pull/238699, this makes rendering of characters a lot closer to how the DOM and in general flow a lot better. It does this by rasterizing glyphs at 0.0, 0.1, ..., 0.9 offsets. Notice how these `a`'s no longer look exactly the same:
    ![Image](https://github.com/user-attachments/assets/08b28943-12fb-486a-9b60-0ece465972da)
- The horizontal scroll bar should now be sized by the GPU lines https://github.com/microsoft/vscode/pull/238679, thought there are still some remaining issues here https://github.com/microsoft/vscode/issues/233790

**Comment by aurexav** [comment]
> [@aurexav](https://github.com/aurexav) thanks, I created [#238849](https://github.com/microsoft/vscode/issues/238849) for 1. For 2, could you create an issue with the code that caused that and what you did if anything to make it show? I'm not too familiar with Rust.

I’m currently unable to reproduce the issue.

I’m encountering a critical bug. When I attempt to open certain files within a Rust project, the editor crashes.

<img width="337" alt="Image" src="https://github.com/user-attachments/assets/33370b24-d23a-4f9e-8f3c-37cf680f5433" />

**Comment by dougg0k** [comment]
I tested `"editor.experimentalGpuAcceleration": "on"` on the 1.97, but all text disappeared, they appeared if I selected them individually.

It was due to webgpu being disabled. So, by adding / using these two flags to `code-flags.conf`, fixed the problem.

```
--enable-unsafe-webgpu
--enable-features=Vulkan
``` 

Even though it gives some warns, it works. Confirmed with `code --status`.

**Comment by pprchal** [comment]
I apologize for the possible offtopic.
But the previous article addresses the problem mentioned here:
https://github.com/Microsoft/vscode/issues/7770

I have a home PC-celeron on a 4K Benq 3840x2560.
If I run vscode by default, then it is unusable.

Those 2 options
```
--enable-unsafe-webgpu
--enable-features=Vulkan
```

are OK, and the code can be used normally (without `--disable-gpu`)

**Comment by mubaidr** [comment]
> I tested `"editor.experimentalGpuAcceleration": "on"` on the 1.97, but all text disappeared, they appeared if I selected them individually.
> 
> It was due to webgpu being disabled. So, by using / adding these two flags to `~/.config/code-flags.conf`, fixed the problem.
> 
> ```
> --enable-unsafe-webgpu
> --enable-features=Vulkan
> ```
> 
> ps: any other feature you try to enable, must be with a comma in the same line, a new enable-features line would overwrite the previous.
> 
> Even though it gives some warns, it works. Confirmed with `code --status`.
> 
> I am on Arch Linux based distro.
> 
> Edit1:
> 
> One issue I found after, once I open a file and start selecting text randomly fast, it lags, not vscode, but the selection. After a bit, it stabilizes, till I open another file and do the same. And same thing with the minimap if you have it enabled.
> 
> Edit2:
> 
> After, I added these other flags to check, I dont which, but at least one helped with the issue.
> 
> ```
> --enable-gpu-rasterization
> --enable-zero-copy
> ```
> 
> It can happen in bigger files, but stops much faster, then it's all fast.
> 
> Edit3:
> 
> I did more testing, I also tested other flags not mentioned here. Besides the slow effects mentioned already, when it's happening if you select with the keyboard word per word or hover with mouse the files tree, all are slower, till all stops being slow all at once. It's one thing for everything.
> 
> Sometimes, when you open the editor, the slow effect are there. Sometimes after opening, all are very fast, but after a while bit the slow appears.
> 
> If you are able to fix the issue, overall it would be an improvement.
> 
> Edit4:
> 
> I will share all the flags I used in attempt to bring an improvement to the issue.
> 
> That slow effect are still random, be on big or small file. Sometimes only on the minimap, sometimes on other parts too. But again, once it stop in one part, it seem to stop in others too. Though it can come back after.
> 
> ```
> --enable-features=Vulkan,CanvasOopRasterization,RawDraw,UseSkiaRenderer
> --enable-unsafe-webgpu
> --enable-gpu-rasterization
> --enable-zero-copy
> --high-efficiency-mode-available
> --scheduler-configuration
> --canvas-oop-rasterization
> --enable-drdc
> --enable-native-gpu-memory-buffers
> --ignore-gpu-blocklist
> --disable-wayland-ime
> --enable-threaded-texture-mailboxes
> --use-vulkan
> --enable-threaded-compositing
> --enable-threaded-animation
> --enable-threaded-scrolling
> --wm-window-animations-disabled
> --animation-duration-scale=0
> --disable-login-animations
> --disable-modal-animations
> --ui-enable-zero-copy
> --enable-raw-draw
> --enable-accelerated-2d-canvas
> --use-skia-renderer
> --disable-gpu-driver-bug-workarounds
> ```
> 
> Update: I added more flags in the list, I dont see the slow happening anymore, I wonder if the skia renderer fixed it, or at least minimized the effect even more.

Confirmed. Enabling webgpu using flag along with Vulkan solves the empy screen issue.

**Comment by kunkhan1994** [comment]
https://github.com/microsoft/vscode/issues/221145

**Comment by dougg0k** [comment]
There is an issue that seem related to GPU accel, that can happen in all places in vscode, it would be nice to have it fixed when this release.

That is, random text overlap, I've seen in editor, and i've seen others reporting happening in the file tree and terminal.

https://github.com/microsoft/vscode/issues/173335
https://github.com/microsoft/vscode/issues/174721

**Comment by larsgeb** [comment]
Just a question from someone who is just getting into the internals of VSCode, and thinks his issue is related.

I have a Jupyter Widget that uses 3D rendering with [PyThreeJS](https://github.com/jupyter-widgets/pythreejs) (and then of course [Three.JS](https://threejs.org/)). This is an interface to WebGL.

On my current VSCode 1.97.2 (Apple Silicon, [this commit](e54c774e0add60467559eb0d1e229c6452cf8447)) and since about summer 2024 (though I wouldn't able to pinpoint a version number), this widget stopped working when executing notebooks directly in VSCode (though they work when launching a kernel from VSCode and going to Safari/FireFox/Chrome). I had a bit of a dive into it recently and found that when the widget gets created, it fails while trying to create a WebGL context.

I checked out a few releases (I think as early as October 2024) of VSCode over the past few months, and MacOS flavours (Intel, Universal), but no joy. I also don't want to regress too far just for this widget, as I'm a happy user of new features.

VSCode's GPU Info shows specifically the following:

```
*   WebGPU has been disabled via blocklist or the command line.
    Disabled Features: webgpu
*   WebGL2 has been disabled via blocklist or the command line.
    Disabled Features: webgl2
*   Accelerated rasterization has been disabled, either via blocklist, about:flags or the command line.
    Disabled Features: rasterization
*   Accelerated video encode has been disabled, either via blocklist, about:flags or the command line.
    Disabled Features: video_encode
*   Accelerated video decode has been disabled, either via blocklist, about:flags or the command line.
    Disabled Features: video_decode
*   WebGL has been disabled via blocklist or the command line.
    Disabled Features: webgl
*   Gpu compositing has been disabled, either via blocklist, about:flags or the command line. The browser will fall back to software compositing and hardware acceleration will be unavailable.
    Disabled Features: gpu_compositing
*   Accelerated 2D canvas is unavailable: either disabled via blocklist or the command line.
    Disabled Features: 2d_canvas
```

Additionally, I've tried most of the Electron command line flags to try and enable this, but also no succes. I think I am not sure what to expect on enabling WebGL at all for version 1.97.2. It seems that on the Insiders Release (1.98.0-insider), WebGL is at least supporter.

I guess this is expected then right? I'm not knowledgeable on the difference between terminal versus editor GPU rendering, so I am also not sure where the "Interactive - ..." window for Jupytext files or the Jupyter Notebook editor directly falls, but in general, on my MacOS Apple Silicon system, this is what VSCode Developer Tools tell me after trying to get a WebGL context.

<img width="547" alt="Image" src="https://github.com/user-attachments/assets/a64c189d-d26d-478a-9bd1-3c2157e857b7" />

Just verifying this is directly related to this issue, otherwise happy to document and write up in a new issue.

**Comment by Tyriar** [comment]
@dougg0k random overlaps in the tree is not related to editor GPU work, I saw another issue related to this recently. The terminal one was a regression https://github.com/microsoft/vscode/issues/239917.

**Comment by Tyriar** [comment]
@larsgeb Chromium maintains a blacklist of hardware to disable webgl on and you might just have a setup that hits that. You could create a new issue and ping @deepak1556 on it since it's not directly related to this one.

**Comment by dougg0k** [comment]
From my testings on the gpu acceleration. Unless you can make webgpu as responsive as with GPU disabled, I dont really see the point in using it.

I included the information in my first comment here.

@Tyriar 

---

In fact, could you tell the actual benefits of using gpu acceleration in a text editor other than taking load off the cpu?

**Comment by Tyriar** [comment]
> From my testings on the gpu acceleration. Unless you can make webgpu as responsive as with GPU disabled, I dont really see the point in using it.

@dougg0k of course, this is a WIP and the focus is on feature parity not performance currently.

> In fact, could you tell the actual benefits of using gpu acceleration in a text editor other than taking load off the cpu?

The goal is to minimize input latency and scroll latency by reduce the time it takes to render a frame as much as possible. This is mostly done by removing the need for browser-based layouts and dealing with DOM nodes which are expensive on the CPU side.

**Comment by axetroy** [comment]
When the layout changes, re-rendering can cause page content to flicker

https://github.com/user-attachments/assets/8b2f10c8-7709-4ffe-bded-eeaf8306b8f6

**Comment by Tyriar** [comment]
Update on this: I was busy with terminal intellisense last month ([which is now in preview!](https://code.visualstudio.com/updates/v1_98#_terminal-intellisense-preview)) and will be mostly improving terminal in copilot reliability and internal security stuff this month.

**Comment by yennhi2906** [comment]
> [@dougg0k](https://github.com/dougg0k) random overlaps in the tree is not related to editor GPU work, I saw another issue related to this recently. The terminal one was a regression [#239917](https://github.com/microsoft/vscode/issues/239917).

**Comment by pinkyo** [comment]
This feature is still deferred this month?

**Comment by aurexav** [comment]
I would like know the roadmap or outline for the next steps. Thanks. @Tyriar 

Is it currently looking like this will be postponed indefinitely?
